workspaces/src/types.rs

```diff
@@ -0,0 +1,243 @@
+        id.0.into_string()
+            .try_into()
+            .expect("Should not fail since this is already validated before")
+    }
+}
+
+impl FromStr for AccountId {
+    type Err = ParseAccountError;
+
+    fn from_str(account_id: &str) -> Result<Self, Self::Err> {
+        Self::validate(account_id)?;
+        Ok(Self(account_id.into()))
+    }
+}
+
+impl fmt::Display for AccountId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// An error occurred when parsing an invalid Account ID with [`AccountId::validate`](crate::AccountId::validate).
+#[derive(Eq, Hash, Clone, Debug, PartialEq)]
+pub struct ParseAccountError(pub(crate) ParseErrorKind, pub(crate) String);
+
+/// A list of errors that occur when parsing an invalid Account ID.
+#[derive(Eq, Hash, Clone, Debug, PartialEq)]
+pub enum ParseErrorKind {
+    TooLong,
+    TooShort,
+    Invalid,
+}
+
+/// Signer that keeps secret key in memory.
+#[derive(Clone)]
+pub struct InMemorySigner {
+    pub account_id: AccountId,
+    pub public_key: PublicKey,
+    pub secret_key: SecretKey,
+}
+
+impl InMemorySigner {
+    pub(crate) fn from_seed(account_id: AccountId, key_type: KeyType, seed: &str) -> Self {
+        let secret_key = SecretKey::from_seed(key_type, seed);
+        Self {
+            account_id,
+            public_key: secret_key.public_key(),
+            secret_key,
+        }
+    }
+
+    pub fn from_secret_key(account_id: AccountId, secret_key: &str) -> Self {
+        let secret_key = SecretKey::from_str(secret_key).expect("Invalid secret key");
+
+        Self {
+            account_id,
+            public_key: secret_key.public_key(),
+            secret_key,
+        }
+    }
+
+    pub fn from_file(path: &Path) -> Self {
+        KeyFile::from_file(path).into()
+    }
+}
+
+impl Signer for InMemorySigner {
+    fn public_key(&self) -> PublicKey {
+        self.public_key.clone()
+    }
+
+    fn sign(&self, data: &[u8]) -> Signature {
+        self.secret_key.sign(data)
+    }
+
+    fn compute_vrf_with_proof(&self, data: &[u8]) -> (Value, Proof) {
+        let secret_key = convert_secret_key(self.secret_key.unwrap_as_ed25519());
+        secret_key.compute_vrf_with_proof(&data)
+    }
+
+    fn write_to_file(&self, path: &Path) {
+        KeyFile::from(self).write_to_file(path);
+    }
+}
+
+impl From<&InMemorySigner> for KeyFile {
+    fn from(signer: &InMemorySigner) -> KeyFile {
+        KeyFile {
+            account_id: signer.account_id.clone(),
+            public_key: signer.public_key.clone(),
+            secret_key: signer.secret_key.clone(),
+        }
+    }
+}
+
+impl From<KeyFile> for InMemorySigner {
+    fn from(key_file: KeyFile) -> Self {
+        Self {
+            account_id: key_file.account_id,
+            public_key: key_file.public_key,
+            secret_key: key_file.secret_key,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct KeyFile {
```

Comment:
Same here: do we need to expose this as an extra type in the API surface, or would something like 

```rust
impl Testnet {
    pub fn from_key_file(path: &Path) -> Self
}
```

would be enough?

Reply:
Ahh yeah, this seems to be just a type we don't need to expose at all. Could just be `pub(crate)` or could just eliminate it completely and move all the functions into the impl of `Signer` and `InMemorySigner` where it gets used